ext/odbc/php_odbc.c

```diff
@@ -2046,35 +2084,39 @@ PHP_FUNCTION(odbc_pconnect)
 /* }}} */
 
 /* {{{ odbc_sqlconnect */
-int odbc_sqlconnect(odbc_connection **conn, char *db, char *uid, char *pwd, int cur_opt, int persistent)
+bool odbc_sqlconnect(zval *zv, char *db, char *uid, char *pwd, int cur_opt, int persistent, zend_string *hash)
 {
 	RETCODE rc;
+	odbc_link *link;
+
+	object_init_ex(zv, odbc_connection_ce);
+	link = Z_ODBC_LINK_P(zv);
+	link->connection = (odbc_connection *)pemalloc(sizeof(odbc_connection), persistent);
+	memset(link->connection, 0, sizeof(odbc_connection));
 
-	*conn = (odbc_connection *)pemalloc(sizeof(odbc_connection), persistent);
-	memset(*conn, 0, sizeof(odbc_connection));
-	(*conn)->persistent = persistent;
-	SQLAllocEnv(&((*conn)->henv));
-	SQLAllocConnect((*conn)->henv, &((*conn)->hdbc));
+	link->persistent = persistent;
+	link->hash = hash;
+	SQLAllocEnv(link->connection->henv);
+	SQLAllocConnect(link->connection->henv, link->connection->hdbc);
 
 #if defined(HAVE_SOLID) || defined(HAVE_SOLID_30)
-	SQLSetConnectOption((*conn)->hdbc, SQL_TRANSLATE_OPTION,
+	SQLSetConnectOption((link->connection->hdbc, SQL_TRANSLATE_OPTION,
 			SQL_SOLID_XLATOPT_NOCNV);
 #endif
 #ifdef HAVE_OPENLINK
 	{
 		char dsnbuf[1024];
 		short dsnbuflen;
 
-		rc = SQLDriverConnect((*conn)->hdbc, NULL, db, SQL_NTS,	dsnbuf, sizeof(dsnbuf) - 1, &dsnbuflen, SQL_DRIVER_NOPROMPT);
+		rc = SQLDriverConnect(link->connection->hdbc, NULL, db, SQL_NTS,	dsnbuf, sizeof(dsnbuf) - 1, &dsnbuflen, SQL_DRIVER_NOPROMPT);
 	}
 #else
 	if (cur_opt != SQL_CUR_DEFAULT) {
-		rc = SQLSetConnectOption((*conn)->hdbc, SQL_ODBC_CURSORS, cur_opt);
+		rc = SQLSetConnectOption(link->connection->hdbc, SQL_ODBC_CURSORS, cur_opt);
 		if (rc != SQL_SUCCESS) {  /* && rc != SQL_SUCCESS_WITH_INFO ? */
-			odbc_sql_error(*conn, SQL_NULL_HSTMT, "SQLSetConnectOption");
-			SQLFreeConnect((*conn)->hdbc);
-			pefree(*conn, persistent);
-			return FALSE;
+			odbc_sql_error(link->connection, SQL_NULL_HSTMT, "SQLSetConnectOption");
+			SQLFreeConnect(link->connection->hdbc);
+			return false;
```

Comment:
Shouldn't we be freeing this here too? Or is this handled somewhere else?

Reply:
Do you mean the removed `pefree(*conn, persistent);` call, right? I believe the link is freed on the call-site of `odbc_sqlconnect()` when `zval_ptr_dtor(return_value);` triggers the free object handler.

Reply:
This probably means that `SQLFreeConnect(link->connection->hdbc);` here should be eliminated...